### PR TITLE
Fix functional tests with CCD enabled

### PR DESCRIPTION
**Pull request by dharmendrak**

### CODE REVIEW

File: src/aat/java/uk/gov/hmcts/cmc/claimstore/tests/functional/RespondToClaimTest.java [hunk 1]
```diff
@@ -44,6 +45,7 @@ public void shouldBeAbleToSuccessfullySubmitAlreadyPaidDefence() {
     }
 
     @Test
+    @Ignore("Full Admission is not implemented, please enable once done")
```
Comment (sabahirfan) [inline src/aat/java/uk/gov/hmcts/cmc/claimstore/tests/functional/RespondToClaimTest.java]: This test passes for CCD disabled, and @geek-soft advised to have it and accepted the risk of functional tests failure for temporarily on env where we have CCD enabled.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/repositories/CCDCaseApi.java [hunk 2]
```diff
@@ -254,6 +269,10 @@ private Claim readCase(User user, String caseId) {
         return search(user, searchString, 1, new ArrayList<>(), null, CaseState.OPEN);
     }
 
+    private List<CaseDetails> searchOnlyClosedCases(User user, Map<String, String> searchString) {
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/repositories/CCDCaseApi.java]: could we have `searchByCaseState(..., CaseState state)` and pass `CaseState` as argument instead of having two different methods?

File: src/main/java/uk/gov/hmcts/cmc/claimstore/repositories/CCDCaseApi.java [hunk 3]
```diff
@@ -118,6 +119,20 @@ public CCDCaseApi(
...
+
+        List<CaseDetails> result = searchOnlyClosedCases(user, ImmutableMap.of("case.externalId", externalId));
+
+        List<Claim> claims = extractClaims(result);
+
+        if (claims.size() > 1) {
+            throw new CoreCaseDataStoreException("More than one claim found by externalId " + externalId);
+        }
+
+        return claims.isEmpty() ? Optional.empty() : Optional.of(claims.get(0));
```
Comment (andrewwa-kainos) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/repositories/CCDCaseApi.java]: simpler as `return claims.stream().findAny();`

File: src/aat/java/uk/gov/hmcts/cmc/claimstore/tests/functional/RespondToClaimTest.java (hunk 1 above)
Comment (andrewwa-kainos) [inline src/aat/java/uk/gov/hmcts/cmc/claimstore/tests/functional/RespondToClaimTest.java]: Can you add a comment about this on the relevant JIRA story?

File: src/main/java/uk/gov/hmcts/cmc/claimstore/repositories/CaseRepository.java [hunk 4]
```diff
@@ -15,6 +15,8 @@
 
     Optional<Claim> getClaimByExternalId(String externalId, String authorisation);
 
+    Optional<Claim> getClosedCaseByExternalId(String externalId, String authorisation);
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/repositories/CaseRepository.java]: Would prefer to have something like `Optional<Claim> getCase(String externalId, String authorisation, CaseState caseState)` instead of `getClosedCaseByExternalId(..)`

File: src/main/java/uk/gov/hmcts/cmc/claimstore/repositories/CCDCaseApi.java [hunk 5]
```diff
@@ -115,7 +102,21 @@ public CCDCaseApi(
             throw new CoreCaseDataStoreException("More than one claim found by externalId " + externalId);
         }
 
-        return claims.isEmpty() ? Optional.empty() : Optional.of(claims.get(0));
+        return claims.stream().findAny();
+    }
+
+    public Optional<Claim> getClosedCasesByExternalId(String externalId, String authorisation, CaseState caseState) {
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/repositories/CCDCaseApi.java]: please rename this method, as it returns cases with other states as well.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ClaimService.java [hunk 6]
```diff
@@ -97,6 +98,12 @@ public Claim getClaimByExternalId(String externalId, String authorisation) {
             .orElseThrow(() -> new NotFoundException("Claim not found by external id " + externalId));
     }
 
+    public Claim getClosedClaimByExternalId(String externalId, String authorisation) {
+        return caseRepository
+            .getCase(externalId, authorisation, CaseState.CLOSED)
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/ClaimService.java]: would prefer to pass the state in parameter and rename method accordingly.

**Comment by geek-soft** [comment]
@dharmendrak, @andrewwa-kainos can you put some description for the PR so that I can understand reason behind these changes?

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ClaimService.java [hunk 7]
```diff
@@ -91,7 +93,7 @@ public Claim getClaimByLetterHolderId(String id, String authorisation) {
             .orElseThrow(() -> new NotFoundException("Claim not found for letter holder id " + id));
     }
 
-    public Claim getClaimByExternalId(String externalId, String authorisation) {
+    public Claim getClaimByExternalId(String externalId, String authorisation, CaseState caseState) {
```
Comment (timja-kainos) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/ClaimService.java]: Why has this parameter been added? It seems to be set to open everywhere. Being open or on hold shouldn’t be a concern at this level

File: src/aat/java/uk/gov/hmcts/cmc/claimstore/tests/functional/SealedClaimPdfTest.java [hunk 8]
```diff
@@ -68,9 +69,14 @@ private String getFullAddressString(Address address) {
     }
 
     private Claim createCase() {
+        UUID externalId = UUID.randomUUID();
```
Comment (geek-soft) [inline src/aat/java/uk/gov/hmcts/cmc/claimstore/tests/functional/SealedClaimPdfTest.java]: No need to create it manually, `testData.submittedByClaimantBuilder()` is generating unique UUID, just do 

```
ommonOperations.submitPrePaymentClaim(claimData.getExternalId().toString(), claimant.getAuthorisation());
```

File: src/aat/java/uk/gov/hmcts/cmc/claimstore/tests/functional/SubmitClaimTest.java [hunk 9]
```diff
@@ -35,9 +35,13 @@ public void before() {
 
     @Test
     public void shouldSuccessfullySubmitClaimDataAndReturnCreatedCase() {
+        UUID externalId = UUID.randomUUID();
```
Comment (geek-soft) [inline src/aat/java/uk/gov/hmcts/cmc/claimstore/tests/functional/SubmitClaimTest.java]: The same here.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/OffersService.java [hunk 10]
```diff
@@ -97,7 +99,7 @@ public Claim countersign(Claim claim, MadeBy party, String authorisation) {
         settlement.countersign(party);
 
         caseRepository.reachSettlementAgreement(claim, settlement, authorisation, SETTLED_PRE_JUDGMENT.name());
-        Claim updated = claimService.getClaimByExternalId(claim.getExternalId(), authorisation);
+        Claim updated = claimService.getClaimByExternalId(claim.getExternalId(), authorisation, CLOSED);
```
Comment (geek-soft) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/OffersService.java]: Have you checked with BA whether closed class should be visible to users dashboards? I cannot recall any requirement to remove settled cases from dashboard.

This seems to be hiding problem, not fixing it.

Comment (timja-kainos) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/OffersService.java]: Do we actually close cases at settlement? I didn't think we had implemented that

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ClaimService.java (hunk 7 above)
Comment (geek-soft) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/ClaimService.java]: As far as I know when case is settled it changes state to 'closed' which:
1) breaks the app as backend is unable to retrieve closed case before producing event
2) removes closed cases from any data retrieval that will result settled case not be displayed on the dashboard / available to users anymore

@dharmendrak as far as I understand fixes #1 but not #2. I find this solution incorrect.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/OffersService.java (hunk 10 above)
Comment (geek-soft) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/OffersService.java]: screen shot 2018-06-15 at 11 29 44

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ClaimService.java (hunk 7 above)
Comment (dharmendrak) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/ClaimService.java]: it was left in error, was not used

File: src/aat/java/uk/gov/hmcts/cmc/claimstore/tests/functional/SubmitClaimTest.java (hunk 9 above)
Comment (dharmendrak) [reply src/aat/java/uk/gov/hmcts/cmc/claimstore/tests/functional/SubmitClaimTest.java]: done

File: src/aat/java/uk/gov/hmcts/cmc/claimstore/tests/functional/SealedClaimPdfTest.java (hunk 8 above)
Comment (dharmendrak) [reply src/aat/java/uk/gov/hmcts/cmc/claimstore/tests/functional/SealedClaimPdfTest.java]: done

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/OffersService.java (hunk 10 above)
Comment (dharmendrak) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/OffersService.java]: it was never a requirement to not show closed state case following claim store model.

File: src/aat/resources/environment.properties [hunk 11]
```diff
@@ -1,6 +1,6 @@
 pdf_service.url = http://dummy-url
 document_management.api_gateway.url = http://dummy-url
-core_case_data.api.url = false
+core_case_data.api.url = http://localhost:4452
```
Comment (geek-soft) [inline src/aat/resources/environment.properties]: Why is that change needed @dharmendrak?

File: src/main/java/uk/gov/hmcts/cmc/claimstore/repositories/CCDCaseApi.java [hunk 12]
```diff
@@ -83,39 +70,17 @@ public CCDCaseApi(
 
     public List<Claim> getBySubmitterId(String submitterId, String authorisation) {
         User user = userService.getUser(authorisation);
-        return extractClaims(searchOnlyOpenCases(user, ImmutableMap.of("case.submitterId", submitterId)));
+        List<CaseDetails> result = searchAll(user, ImmutableMap.of("case.submitterId", submitterId));
+        List<CaseDetails> validCases = result.stream().filter(c -> !isCaseOnHold(c)).collect(Collectors.toList());
```
Comment (geek-soft) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/repositories/CCDCaseApi.java]: Instead of creating two arrays I would inline it to be:

```
searchAll(user, ImmutableMap.of("case.submitterId", submitterId)).stream()
  .filter(c -> !isCaseOnHold(c))
  .collect(Collectors.toList())
```

File: src/main/java/uk/gov/hmcts/cmc/claimstore/repositories/CCDCaseApi.java [hunk 13]
```diff
@@ -163,6 +128,24 @@ public void linkDefendant(String authorisation) {
             ).forEach(caseId -> linkToCase(defendantUser, anonymousCaseWorker, letterHolderId, caseId)));
     }
 
+    private Optional<Claim> getCaseBy(String input, String authorisation, ImmutableMap<String, String> searchString) {
```
Comment (geek-soft) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/repositories/CCDCaseApi.java]: Please remove `input` since it duplicates value of `ImmutableMap<String, String> searchString`. Also no point to pass argument just for logging purpose. Either print search criteria in logs or case ID retrieved from data store.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/repositories/CCDCaseApi.java [hunk 14]
```diff
@@ -128,19 +132,19 @@ public void linkDefendant(String authorisation) {
...
         if (result.size() == 1 && isCaseOnHold(result.get(0))) {
-            throw new OnHoldClaimAccessAttemptException("Case is on hold " + input);
+            return Optional.empty();
         }
 
         List<Claim> claims = extractClaims(result);
 
         if (claims.size() > 1) {
-            throw new CoreCaseDataStoreException("More than one claim found by input " + input);
+            throw new CoreCaseDataStoreException("More than one claim found by search String " + searchString);
```
Comment (geek-soft) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/repositories/CCDCaseApi.java]: Please lower case `String`.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/repositories/CCDCaseApi.java [hunk 15]
```diff
@@ -163,6 +132,24 @@ public void linkDefendant(String authorisation) {
...
+        List<CaseDetails> result = searchAll(user, searchString);
+
+        if (result.size() == 1 && isCaseOnHold(result.get(0))) {
+            return Optional.empty();
+        }
+
+        List<Claim> claims = extractClaims(result);
+
+        if (claims.size() > 1) {
+            throw new CoreCaseDataStoreException("More than one claim found by search String " + searchString);
```
Comment (geek-soft) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/repositories/CCDCaseApi.java]: Please lowercase `String` here.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/repositories/CCDCaseApi.java [hunk 16]
```diff
@@ -31,25 +32,11 @@
 import java.util.Optional;
 import java.util.stream.Collectors;
 
+import static uk.gov.hmcts.cmc.ccd.domain.CaseState.OPEN;
+
 @Service
 @ConditionalOnProperty(prefix = "core_case_data", name = "api.url")
 public class CCDCaseApi {
-
-    public static enum CaseState {
```
Comment (timja-kainos) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/repositories/CCDCaseApi.java]: Does this still need to be moved from here?
